outlines/caching.py

```diff
@@ -1,20 +1,74 @@
+import asyncio
+import hashlib
 import os
 from typing import Callable, Optional
 
-from perscache import Cache, NoCache
-from perscache.serializers import JSONSerializer
-from perscache.storage import LocalFileStorage
+import cloudpickle
+from diskcache import Cache
 
 home_dir = os.path.expanduser("~")
 cache_dir = os.environ.get("OUTLINES_CACHE_DIR", f"{home_dir}/.cache/outlines")
-memory = Cache(serializer=JSONSerializer(), storage=LocalFileStorage(cache_dir))
-
-
-def cache(ignore: Optional[str] = None):
-    def cache_fn(fn: Callable):
-        return memory.cache(ignore=ignore)(fn)
+memory = Cache(cache_dir, eviction_policy="none", cull_limit=0)
+_caching_enabled = True
+
+
+def hash_arguments(*args, **kwargs) -> str:
+    """Create a hash out of the args and kwargs provided"""
+    result = hashlib.md5()
+    for item in list(args) + sorted(kwargs.items()):
+        result.update(cloudpickle.dumps(item))
+    return result.hexdigest()
+
+
+def cache(key_function: Optional[Callable] = None):
+    """Caching decorator for memoizing function calls.
+    The cache key is created based on the values returned by the key_function callable
+    if provided or based on the arguments of the decorated function directly otherwise
+    Parameters
+    ----------
+    key_function
+      A callable function used to generate a unique key for each function call. It's
+      called with the arguments of the decorated function as arguments
+    Returns
+    -------
```

Comment:
```suggestion
    Parameters
    ----------
    
    key_function
      A callable function used to generate a unique key for each function call. It's
      called with the arguments of the decorated function as arguments
      
    Returns
    -------
```

Reply:
Also, does will this work with the OpenAI integration?

Reply:
Yes. I've added `async_wrapper` for it to work with the OpenAI integration